### PR TITLE
Inconsistent date parsing of to_datetime

### LINKED ISSUE #12585: Inconsistent date parsing of to_datetime
As this comes up regularly:
- [x] https://github.com/pydata/pandas/issues/12583
- [x] https://github.com/pydata/pandas/issues/12501
- [x] https://github.com/pydata/pandas/issues/7348
- [x] https://github.com/pandas-dev/pandas/issues/15075
- [x] #12611 deal with `yearfirst/datefirst` in `read_csv`
- [x] #12730, compat with dateutil 2.5.2 `dayfirst=True, yearfirst=True` fixes
- [x] https://github.com/pydata/pandas/issues/14301 raise if inconsistency in `read_csv` with `dayfirst/yearfirst`
- [x] performance https://github.com/pandas-dev/pandas/issues/25848
**The problem**

Depending on the string formats, you can get inconsistent parsing (meaning: different interpretation of parts of the strings as days vs months between different values in the passed list/array of strings). 
For example:

```
In [22]: pd.to_datetime(["31/12/2014", "10/03/2011"])
Out[22]: DatetimeIndex(['2014-12-31', '2011-10-03'])
```

where the first is parsed as 'dayfirst' and the second not. Above particular case can be solved by passing `dayfirst=True`, but in any case, the default behaviour of `to_datetime` results in inconsistent parsing here. 
Another typical case if where you do provide `dayfirst=True` but have a malformed date, then it will fall back to 'not dayfirst' (https://github.com/pydata/pandas/issues/3341).

**The reason**

The reason is the flexibility of `dateutil.parser.parse`. This is in the first place also a feature, as it lets you parse messy datetime colums. But when it comes to flexibility regarding `dayfirst`/`yearfirst`, this gives unexpected results. 
So dateutil will 'try' to use `dayfirst=True`, but when this does not give a valid date, it will ignore `dayfirst=True` and parse the string anyway:

```
In [3]: dateutil.parser.parse('1/12/2012', dayfirst=True)
Out[3]: datetime.datetime(2012, 12, 1, 0, 0)

In [4]: dateutil.parser.parse('1/13/2012', dayfirst=True)
Out[4]: datetime.datetime(2012, 1, 13, 0, 0)
```

The same is true for `dayfirst=False` (the default) as this will also try with 'day first' if the other way around failed. 
The issue for the lack of strictness of dayfirst is actually https://github.com/pydata/pandas/issues/3341.

**Possible solutions**
- In some cases passing `dayfirst=True` is sufficient (if you don't have messy or malformed dates).
- The typical answer is saying: "if you want to be sure to have strict and consistent parsing, provide a `format` argument. 
  We should probably stress this more in the docstring of `to_datetime` 

But, regardless of the two points above, fact is that the _default_ behaviour of `to_datetime` (without passing any arguments) can do inconsistent parsing and so give unexpected results. 
Possible ways to have better default behaviour:
- Have the possibility to specify that `dateutil` should be strict about `dayfirst`. I raised an issue for that here: https://github.com/dateutil/dateutil/issues/214
- Changing the default of `infer_datetime_format` to `True` would solve it in some cases (where the first datetime has the correct format you want), but certainly not all.
- The idea of raising a warning when inconsistent parsing happened, is AFAIK not possible as pandas does not know how dateutil parsed the dates. And always warning when no format is passed or could be inferred (so in practice, when dateutil is used), is maybe to drastically, as the flexible parsing is also a feature.

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] xref #12585
- [x] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

carries on from https://github.com/pandas-dev/pandas/pull/35428

### CODE REVIEW

**Comment by jreback** [comment]
@MarcoGorelli this looks neat. prob need to troll all of the issues and add every test you can find :->

**Review by mroeschke — COMMENTED** [review]
Nice looks good. Just needs a whatsnew note for 1.4 (probably good to have a dedicated section demoing this behavior), and maybe the `to_datetime` docstring can explain when a warning will be raised.

**Comment by MarcoGorelli** [comment]
Sure, added, here's the docstring:

image

Unfortunately, this doesn't solve the case of datetime strings, just date strings. E.g.:
```python
In [2]: pd.to_datetime(['31-12-2021'])
<ipython-input-2-a6d9926683b2>:1: UserWarning: Parsing '31-12-2021' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  pd.to_datetime(['31-12-2021'])
Out[2]: DatetimeIndex(['2021-12-31'], dtype='datetime64[ns]', freq=None)

In [3]: pd.to_datetime(['31-12-2021 12:00:31'])
Out[3]: DatetimeIndex(['2021-12-31 12:00:31'], dtype='datetime64[ns]', freq=None)
```
because those cases go straight to dateutil:

https://github.com/pandas-dev/pandas/blob/08d296f1278e08b407448c95086589e1d10285f9/pandas/_libs/tslibs/parsing.pyx#L238-L244

and this PR only addresses what reaches `_parse_delimited_date`.

> Just needs a whatsnew note for 1.4 (probably good to have a dedicated section demoing this behavior),

As in, in the "notable bug fixes" section? If so, I've added it there, here's how it looks:

image

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/blob/08d296f1278e08b407448c95086589e1d10285f9/pandas/_libs/tslibs/parsing.pyx#L238-L244

yeah would be nice to warn in this as well

**Comment by MarcoGorelli** [comment]
> yeah would be nice to warn in this as well

Agreed, just not sure how to do it - all I can think of is

```python
        dt = du_parse(date_string, default=_DEFAULT_DATETIME,
                      dayfirst=dayfirst, yearfirst=yearfirst, **kwargs)
        if dayfirst and not re.search(rf'(?<!\d){dt.day}(?!\d)', date_string).start() < re.search(rf'(?<!\d){dt.month}(?!\d)', date_string).start():
            warnings.warn(
                PARSING_WARNING_MSG.format(
                    date_string=date_string,
                    format='MM/DD/YYYY'
                ),
                stacklevel=4,
            )
        elif not dayfirst and re.search(rf'(?<!\d){dt.day}(?!\d)', date_string).start() < re.search(rf'(?<!\d){dt.month}(?!\d)', date_string).start():
            warnings.warn(
                PARSING_WARNING_MSG.format(
                    date_string=date_string,
                    format='DD/MM/YYYY'
                ),
                stacklevel=4,
            )
```
would that be OK?

**Comment by jreback** [comment]
well the bigger problem here is that we *really* should warn anytime this is called `du_parse` (if we are not guessing a format) and actually parsing, e.g. this is the fallback case.

I am happy to merge this and to do followups, but ideally want to warn if at all possible that something is amiss (even with false positives).

File: doc/source/whatsnew/v1.4.0.rst [hunk 1]
```diff
@@ -50,10 +50,19 @@ Notable bug fixes
 
 These are bug fixes that might have notable behavior changes.
 
-.. _whatsnew_140.notable_bug_fixes.notable_bug_fix1:
+.. _whatsnew_140.notable_bug_fixes.inconsistent_date_string_parsing:
 
-notable_bug_fix1
-^^^^^^^^^^^^^^^^
+Inconsistent date string parsing
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:55]: can you also add to the documentation itself, e.g. somewhere in `timeseries.rst` is appropriate

File: pandas/core/tools/datetimes.py [hunk 2]
```diff
@@ -701,8 +701,15 @@ def to_datetime(
         Specify a date parse order if `arg` is str or its list-likes.
         If True, parses dates with the day first, eg 10/11/12 is parsed as
         2012-11-10.
-        Warning: dayfirst=True is not strict, but will prefer to parse
-        with day first (this is a known bug, based on dateutil behavior).
+
+        .. warning::
+
+            dayfirst=True is not strict, but will prefer to parse
+            with day first (this is a known bug, based on dateutil behavior).
```
Comment (jreback) [inline pandas/core/tools/datetimes.py:708]: i wouldn't say this is a known bug unless you can point to an authoritative reference.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
does this show any warnings during csv parsing? (i think it should in the same cases). ideally create a followon issue with checkboxes for the cases.

**Comment by jreback** [comment]
@MarcoGorelli looks pretty good if you can rebase & address comments.

File: pandas/core/tools/datetimes.py (hunk 2 above)
Comment (MarcoGorelli) [reply pandas/core/tools/datetimes.py:708]: I didn't add this, it's from https://github.com/pandas-dev/pandas/pull/7599 , I just put it into a warning block

**Comment by MarcoGorelli** [comment]
> does this show any warnings during csv parsing? (i think it should in the same cases).

yup, I've added the same suite of tests for that

> well the bigger problem here is that we really should warn anytime this is called du_parse (if we are not guessing a format) and actually parsing, e.g. this is the fallback case.

I gave that a go but it's a lot more complicated and there's a ton more warnings to catch. I've pushed that work to a branch (MarcoGorelli:warn-on-du-parse), but for now I've kept this PR to just the delimited date case

**Review by mroeschke — APPROVED** [review]

**Review by mroeschke — COMMENTED** [review]
Just a small comment about removing the comment about the dateutil bug and then feel free to merge.

**Review by mroeschke — APPROVED** [review]

**Pull request merged**